File: helloWord/helloRedis/NodeJs/helloRedis.js
/**
 * Redis HelloWorld for Node.js (written in JavaScript)
 * Could be used in an electron app
 * 
 * See: https://github.com/NodeRedis/node_redis 
 */

// first install the required redis package 
// > npm install redis
var redis = require("redis");

// create Client connection
// can be used as template for other (subscribing) clients
var client = redis.createClient({
  host: "localhost"
  , port: 6379
  , db: 0
});

client.on("error", function (err) {
  console.log("Error " + err);
  client.quit();
});


/** #################################################################
 * GET and SET Redis values #########################################
 ################################################################# */

var val = "helloRedisTest";
red.set("helloRedis/nodejs", val.toString()); // for other values than strings, you should call the .toString() method. Redis can only store strings.

// In nodejs redis all requests are handled with callback functions 
red.get("helloRedis/nodejs", function (err, reply) {
  if (reply) {
    console.log("Got value for helloRedis/nodejs:" + reply.toString());
  }
});



/** #################################################################
 * client management for subscribing and publishing #################
 ################################################################# */

// stores all subscribing clients created with this module
var clientList = {};
// incrementing client id for 
var clientID = 0;

/**
 * Method to create a new client and add this client to the clientList
 */
function getNewClient() {
  // Duplicate client
  var c = client.duplicate();

  // Create object containing the new client an a new  client id
  var retObj = { client: c, id: clientID }
  clientList[clientID++] = c;

  return retObj;
}

/**
 * Close a client with the given id from the client list
 * @param {Number} id client id 
 */
function closeClient(id) {
  var c = clientList[id];
  if (c) {
    c.quit();
    c = null;
    delete clientList[id];
  }
}



/** #################################################################
 * EXPORT objects ###################################################
 ################################################################# */

module.exports = {

  /**
   * @returns a new redis client
   */
  getClient: function () {
    return client.duplicate();
  },

  /**
   * Subscribe to a specific redis key
   * @param {string} key Redis Key of topic you want to subscribe to 
   * @param {Function} callback callback method that is called as soon as a new value is published under the topic
   */
  subscribeKey: function (key, callback) {
    var c = getNewClient();
    var sub = c.client;
    sub.on("error", function (err) {
      console.log("Error " + err);
      closeClient(c.id);
    });

    sub.on("message", function (channel, message) {
      console.log("sub channel " + channel + ": " + message);
      callback(message);
    });

    sub.on("ready", function () {
      sub.subscribe(key);
    })
  },

  /**
   * Publish a value to a channel
   * @param {String} key key to publish to
   * @param {String} value value to be published
   */
  publish: function (key, value) {
    if (client.connected) {
      console.log("Published " + String(value) + " at " + key);
      client.publish(key, String(value));
    }
  },

  /**
   * Function to properly close RedisConnector
   */
  closeAll: function () {

    var keys = [];
    for (const [key, value] of Object.entries(clientList)) {
      keys.push(key);
    }

    keys.forEach(k => {
      closeClient(k);
    });
  }
}

File: helloWord/helloRedis/Python/helloRedis.py
#!/usr/bin/env python3

"""
Python3 HelloWorld for Redis Key-Value-Database

Taken from: https://github.com/andymccurdy/redis-py 

Official Redis Commands:
https://redis.io/commands 

If necessary, install Python redis library:
$ pip install redis

"""

import redis    # redis import
import time     # used for examples

""" Example for setting and getting values """
def normaleExample():
  # Connection to local redis db, default port ist 6379
  r = redis.Redis(host="localhost", port=6379, db=0)

  # set value to key
  r.set("helloRedis/python", "test")

  # get value from key
  r.get("helloRedis/python")


""" Example for pipelining commands.
    Pipelines provide support for buffering multiple commands to the server in a single request. 
    They can be used to dramatically increase the performance of groups of commands by 
    reducing the number of back-and-forth TCP packets between the client and server.  """
def pipelineExample():
  # Connection to local redis db, default port ist 6379
  r = redis.Redis(host="localhost", port=6379, db=0)

  r.set("helloRedis/python", "test")

  # Use the pipeline() method to create a pipeline instance
  pipe = r.pipeline()
  # The following SET commands are buffered
  r.set("helloRedis/python/p1", "test1")

  # calls can be chained like:
  r.set("helloRedis/python/p2", "test2").set("helloRedis/python/p3", "test3")

  pipe.get('helloRedis/python')

  # the EXECUTE call sends all buffered commands to the server, returning
  # a list of responses, one for each command.
  pipe.execute()
  # --> [True, True, True, 'test']


"""
  Example for publishing and subscribing to key.
  Redis-Py implements a PubSub object that subscribes to channels and listens for new messages.
"""
def publishSubscribeExample():
  # Connection to local redis db, default port ist 6379
  r = redis.Redis(host="localhost", port=6379, db=0)

  p = r.pubsub()
  # or to ignore subscribe messages
  p = r.pubsub(ignore_subscribe_messages=True) 


  # Subscribe for specific key
  p.subscribe('helloRedis/python', 'helloRedis/python/p1')
  # or subscribe to a pattern
  p.psubscribe('helloRedis/*')

  # >>> p.get_message()
  # {'pattern': None, 'type': 'subscribe', 'channel': 'helloRedis/python', 'data': 1L}
  # >>> p.get_message()
  # {'pattern': None, 'type': 'subscribe', 'channel': 'helloRedis/python/p1', 'data': 2L}
  # >>> p.get_message()
  # {'pattern': None, 'type': 'psubscribe', 'channel': 'helloRedis/*', 'data': 3L}

  # the publish method returns the number matching channel and pattern
  # subscriptions. 'helloRedis/python' matches both the 'helloRedis/python'
  # subscription and the 'helloRedis/*' pattern subscription, so this message will
  # be delivered to 2 channels/patterns
  r.publish('helloRedis/python', 'some data') # == 2

  # >>> p.get_message()
  # {'channel': 'helloRedis/python', 'data': 'some data', 'pattern': None, 'type': 'message'}
  # >>> p.get_message()
  # {'channel': 'helloRedis/python', 'data': 'some data', 'pattern': 'helloRedis/*', 'type': 'pmessage'}

  ### Callback Method ###
  p.subscribe(**{'helloRedis/python': pubSubHandler})
  r.publish('helloRedis/python', 'awesome data')
  p.get_message() # --> HANDLER: awesome data


  ### to get messages ###
  # 1. polling:
  while True:
    message = p.get_message()
    if message:
      # do something with the message
      pass
    time.sleep(0.001) 
  
  # 2. blocking:
  for message in p.listen():
    # do something with the message
    pass


  # CLose publisher subscriber and unsubscribe from all channels
  p.close()  


  



""" Callback Method for PubSub """
def pubSubHandler(message):
  print("HANDLER: ", message['data'])







